app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarUIView.kt

```diff
@@ -37,56 +44,179 @@ class AwesomeBarUIView(
+            actionEmitter.onNext(AwesomeBarAction.SearchTermsTapped(searchTerms, searchEngine))
         }
+    }
 
+    val shortcutSearchUseCase = object : SearchUseCases.SearchUseCase {
+        override fun invoke(searchTerms: String, searchEngine: SearchEngine?) {
+            actionEmitter.onNext(AwesomeBarAction.SearchTermsTapped(searchTerms, state?.suggestionEngine))
+        }
+    }
+
+    init {
         with(container.context) {
-            view.addProviders(ClipboardSuggestionProvider(
+            clipboardSuggestionProvider = ClipboardSuggestionProvider(
                 this,
                 loadUrlUseCase,
                 getDrawable(R.drawable.ic_link)!!.toBitmap(),
                 getString(R.string.awesomebar_clipboard_title)
                 )
-            )
+
+            sessionProvider =
+                SessionSuggestionProvider(
+                    components.core.sessionManager,
+                    components.useCases.tabsUseCases.selectTab,
+                    components.utils.icons
+                )
+
+            historyStorageProvider =
+                HistoryStorageSuggestionProvider(
+                    components.core.historyStorage,
+                    loadUrlUseCase,
+                    components.utils.icons
+                )
 
             if (Settings.getInstance(container.context).showSearchSuggestions()) {
-                view.addProviders(
+                val draw = getDrawable(R.drawable.ic_search)
+                draw?.setTint(ContextCompat.getColor(this, R.color.search_text))
+
+                defaultSearchSuggestionProvider =
                     SearchSuggestionProvider(
                         searchEngine = components.search.searchEngineManager.getDefaultSearchEngine(this),
                         searchUseCase = searchUseCase,
                         fetchClient = components.core.client,
                         mode = SearchSuggestionProvider.Mode.MULTIPLE_SUGGESTIONS,
                         limit = 3
+                        // TODO icon = draw?.toBitmap()
                     )
-                )
             }
 
-            view.addProviders(
-                SessionSuggestionProvider(
-                    components.core.sessionManager,
-                    components.useCases.tabsUseCases.selectTab,
-                    components.utils.icons
-                ),
-                HistoryStorageSuggestionProvider(
-                    components.core.historyStorage,
-                    loadUrlUseCase,
-                    components.utils.icons
-                )
+            shortcutsEnginePickerProvider =
+                    ShortcutsSuggestionProvider(
+                        components.search.searchEngineManager,
+                        this,
+                        ::selectShortcutEngine,
+                        ::selectShortcutEngineSettings)
+        }
+    }
+
+    private fun showShortcutEnginePicker() {
```

Comment:
Is it bad that all this logic for showing the search shortcut provider / shortcut engine settings exists inside of AwesomeBarUIView? If so, what where you recommend I move it?

Reply:
The class is getting a bit large and doing a lot of things. Worst case, you can always break out related methods to a class of extension methods.